shared/src/commonMain/kotlin/org/hyperskill/app/track_selection/presentation/TrackSelectionListActionDispatcher.kt

```diff
@@ -0,0 +1,100 @@
+import ru.nobird.app.presentation.redux.dispatcher.CoroutineActionDispatcher
+
+class TrackSelectionListActionDispatcher(
+    config: ActionDispatcherOptions,
+    private val analyticInteractor: AnalyticInteractor,
+    private val sentryInteractor: SentryInteractor,
+    private val trackInteractor: TrackInteractor,
+    private val progressesInteractor: ProgressesInteractor,
+    private val profileInteractor: ProfileInteractor,
+    private val currentStudyPlanStateRepository: CurrentStudyPlanStateRepository
+) : CoroutineActionDispatcher<Action, Message>(config.createConfig()) {
+    override suspend fun doSuspendableAction(action: Action) {
+        when (action) {
+            is InternalAction.Initialize ->
+                coroutineScope {
+                    val sentryTransaction =
+                        HyperskillSentryTransactionBuilder.buildTrackSelectionListScreenRemoteDataLoading()
+                    sentryInteractor.startTransaction(sentryTransaction)
+
+                    val studyPlanDeferred = async { currentStudyPlanStateRepository.getState() }
+                    val tracksDeferred = async { trackInteractor.getAllTracks() }
+
+                    val tracks = tracksDeferred.await()
+                        .getOrElse {
+                            sentryInteractor.finishTransaction(sentryTransaction, it)
+                            return@coroutineScope onNewMessage(TrackSelectionListFeature.TracksFetchResult.Error)
+                        }
+                    val trackProgressById = progressesInteractor
+                        .getTracksProgresses(tracks.map { it.id }, forceLoadFromRemote = true)
+                        .getOrElse {
+                            sentryInteractor.finishTransaction(sentryTransaction, it)
+                            return@coroutineScope onNewMessage(TrackSelectionListFeature.TracksFetchResult.Error)
+                        }
+                        .associateBy { it.id }
+
+                    if (tracks.size != trackProgressById.size) {
+                        sentryInteractor.captureErrorMessage("TrackList: tracks.size != tracksProgresses.size")
+                    }
+
+                    val tracksWithProgresses = tracks.mapNotNull { track ->
+                        trackProgressById[track.progressId]?.let { trackProgress ->
+                            TrackWithProgress(
+                                track = track,
+                                trackProgress = trackProgress
+                            )
+                        }
+                    }
+
+                    val currentTrackId = studyPlanDeferred.await()
+                        .getOrElse {
+                            sentryInteractor.finishTransaction(sentryTransaction, it)
+                            return@coroutineScope onNewMessage(TrackSelectionListFeature.TracksFetchResult.Error)
+                        }
+                        .trackId
```

Comment:
@ivan-magda 
In that case, studyPlan fetching would be synchronous.
``` kotlin
val studyPlanDeferred = async { currentStudyPlanStateRepository.getState() }
val currentTrackId = studyPlanDeferred.await()
    .getOrElse {
        sentryInteractor.finishTransaction(sentryTransaction, it)
        return@coroutineScope onNewMessage(TrackSelectionListFeature.TracksFetchResult.Error)
     }
    .trackId
val tracksDeferred = async { trackInteractor.getAllTracks() }
```
I guess it is not so important to wait one more second in case of an error.

Reply:
In future, there we'll use this screen for selection of track by new user, in this case call of `currentStudyPlanStateRepository.getState()` will cause network request, I agree with @ivan-magda

Reply:
@vladkash 
Then, it's better to make the study-plan request synchronous and to add a comment about the reason to make it clear in the future.

Reply:
> Then, it's better to make the study-plan request synchronous and to add a comment about the reason to make it clear in the future.

@XanderZhu see this comment https://github.com/hyperskill/mobile-app/pull/475#discussion_r1193584868 for new users it's better to leave async